### ISSUE TITLE
Kafka consumer does not receive correct offset

### ISSUE THREAD

**Issue by ilyakorol** [ClickHouse/ClickHouse#9334]
**Describe the bug or unexpected behaviour**
My configuration: 
- Kafka cluster, topic with 4 partitions  
- 4 ClickHouse servers with KafkaEngine on each server

Sometimes Kafka consumer just stop receiving the right offset from Kafka (message occasionally becomes the same forever):
```
2020.02.24 04:41:59.931351 [ 36 ] {} <Trace> StorageKafka (topic_name_queue): Polled offset 11677502 (topic: topic_name, partition: 0)
2020.02.24 04:41:59.931392 [ 36 ] {} <Trace> StorageKafka (topic_name_queue): Nothing to commit.
2020.02.24 04:41:59.932957 [ 36 ] {} <Trace> StorageKafka (topic_name_queue): Committed offset 11677502 (topic: topic_name, partition: 0)
2020.02.24 04:41:59.933098 [ 36 ] {} <Trace> StorageKafka (topic_name_queue): Execution took 512 ms.
```
In Kafka itself I can see that offset for partition_0 increases, everything is ok, but in ClickHouse logs offset remains the same as it was a couple of hours ago.
Moreover, other servers and partitions operate perfectly.

Kafka logs are clean, ClickHouse server restart or consumer reattach helps. 

ClickHouse version: 19.16.13.54

**Comment by filimonov** [comment]
Interesting.
1) if you have multiple kafka brokers - please check if they are in sync.
2) Please find in logs the first time that offset appears, and check if something supsupcious was happening that time. 
3) if it possible please enable librdkafka logging by adding the following line to your config.xml.
```xml
  <kafka>
    <debug>all</debug>
  </kafka>
```
After that restart clickhouse server (or do DETACH / ATTACH Engine=Kafka table).
Those logs can be quite noisy and should be written to `/var/log/clickhouse-server/stderr` file (in docker - to docker logs). Please wait till the issue will reproduce and send the fragments of that log.

**Comment by ilyakorol** [comment]
1. It looks like everything ok with brokers (other topics/partitions/consumers work just fine)
2. Already done - nothing criminal neither in Clickhouse logs nor in Kafka logs
3. Will do

Thank you for the quick reply.

**Comment by ilyakorol** [comment]
We have 5 topics with 4 partitions per topic, it seems like only consumers from partition #0 could be broken.

**Comment by filimonov** [comment]
Again that sounds a bit like some for brokers is not in sync.

Please try
```
./kafka-topics.sh --describe --zookeeper ${ip}:2181 --topic test
```

'Isr' number (in-sync-replicas) should be the same as 'replicas'

**Comment by ilyakorol** [comment]
As can I see, everything is okay, at least now:
```
Topic: topic_name    PartitionCount: 4    ReplicationFactor: 2    Configs: flush.ms=5000,segment.bytes=1073741824,flush.messages=5000,max.message.bytes=2000024
    Topic: topic_name    Partition: 0    Leader: 2    Replicas: 2,3    Isr: 2,3
    Topic: topic_name    Partition: 1    Leader: 3    Replicas: 3,0    Isr: 0,3
    Topic: topic_name    Partition: 2    Leader: 0    Replicas: 0,1    Isr: 1,0
    Topic: topic_name    Partition: 3    Leader: 1    Replicas: 1,2    Isr: 2,1
```

**Comment by filimonov** [comment]
Please monitor your Kafka cluster, especially for in-sync replica (ISR) shrinks, under-replicated partitions, and unpreferred leaders. The replica is marked as out-of-sync only after reaching `replica.lag.time.max.ms` or `replica.lag.max.messages` (those are broker settings).

It may be that some broker goes out of the sync, and it may be visible only on some particular partition when that partition, for example, get more data (due to more intensive load, caused by used partitioning schema)

**Comment by filimonov** [comment]
Any updates here?

**Comment by youssefqarri** [comment]
I have the same issue, with a basic configuration :
  - one partition
  - CSV format (not json)

It works for few hours, and then, with any intervention or modification of the messages formats, it stops working.

**Comment by filimonov** [comment]
@youssefqarri can you follow the recommendations above to collect debug info?

**Comment by youssefqarri** [comment]
@filimonov I already tried without any success. I stopped using Kafka Engine because it is not reliable. 

As I am running the functionality in production, I preferred to switch to importing CSV with a custom script using CRON TAB.

By the way, I also tried JSON in first place but import stops working if my logs contains many invalid JSON line.

**Comment by filimonov** [comment]
> @filimonov I already tried without any success.

No success in collecting logs? 

> I stopped using Kafka Engine because it is not reliable.

That says nothing about the real problems you face.

I don't know which version did you use, no logs, that exactly went wrong. No clues giving ideas how to help you or how to improve ClickHouse. 

> By the way, I also tried JSON in first place but import stops working if my logs contains many invalid JSON line.

And what was the problem? You can skop broken messages in Kafka engine if you need.

> I preferred to switch to importing CSV with a custom script using CRON TAB.

Sure, that is a good way. If you don't use Kafka for something else, there is no reason to introduce that for ClickHouse only. Both tools need some expertise, it's not worth to start learning both simulataniosly.

**Comment by radist101** [comment]
I have a similar problem
Docker settings:
```
services:

  clickhouse-server:
    image: yandex/clickhouse-server:latest
    restart: always
    ports:
      - 8123:8123
      - 9000:9000
    ulimits:
      nproc: 65535
      nofile:
       soft: 262144
       hard: 262144

 kafka:
    image: "wurstmeister/kafka:latest"
    ports:
      - "9092:9092"
    hostname: kafka
    depends_on:
      - zookeeper
    environment:
      - KAFKA_ADVERTISED_HOST_NAME=192.168.43.140
      - KAFKA_ADVERTISED_LISTENERS=PLAINTEXT://192.168.43.140:9092
      - KAFKA_ZOOKEEPER_CONNECT=zookeeper:2181

 zookeeper:
    image: wurstmeister/zookeeper:latest
    hostname: zookeeper
    ports:
      - "2181:2181"
```

CH logs contain the following:
```
2020.04.30 13:59:59.211356 [ 61 ] {} <Debug> StorageKafka (event_props_queue): Started streaming to 1 attached views
2020.04.30 13:59:59.212394 [ 61 ] {} <Trace> StorageKafka (event_props_queue): Already subscribed to topics: [ event_topic ]
2020.04.30 13:59:59.212617 [ 61 ] {} <Trace> StorageKafka (event_props_queue): Already assigned to : [ event_topic[0:#] ]
2020.04.30 13:59:59.713468 [ 61 ] {} <Trace> StorageKafka (event_props_queue): Stalled
2020.04.30 13:59:59.713681 [ 61 ] {} <Trace> StorageKafka (event_props_queue): Polled offset 1 (topic: event_topic, partition: 0)
2020.04.30 13:59:59.713707 [ 61 ] {} <Trace> StorageKafka (event_props_queue): Nothing to commit.
2020.04.30 13:59:59.716272 [ 61 ] {} <Trace> StorageKafka (event_props_queue): Committed offset 1 (topic: event_topic, partition: 0)
2020.04.30 13:59:59.716366 [ 61 ] {} <Trace> StorageKafka (event_props_queue): Execution took 505 ms.
```

Data itself falls into Kafka.

In Kafka logs:
```
kafka_1               | [2020-04-30 13:55:37,792] INFO [Partition event_topic-0 broker=1001] No checkpointed highwatermark is found for partition event_topic-0 (kafka.cluster.Partition)
kafka_1               | [2020-04-30 13:55:37,792] INFO [Partition event_topic-0 broker=1001] Log loaded for partition event_topic-0 with initial high watermark 0 (kafka.cluster.Partition)
kafka_1               | [2020-04-30 13:55:37,792] INFO [Partition event_topic-0 broker=1001] event_topic-0 starts at Leader Epoch 0 from offset 0. Previous Leader Epoch was: -1 (kafka.cluster.Partition)
kafka_1               | [2020-04-30 13:55:42,751] INFO [GroupCoordinator 1001]: Preparing to rebalance group group1 in state PreparingRebalance with old generation 0 (__consumer_offsets-40) (reason: Adding new member ClickHouse 20.3.8.53-e52a90de-e972-492f-b853-3ecca7352c76 with group instanceid None) (kafka.coordinator.group.GroupCoordinator)
kafka_1               | [2020-04-30 13:55:42,768] INFO [GroupCoordinator 1001]: Stabilized group group1 generation 1 (__consumer_offsets-40) (kafka.coordinator.group.GroupCoordinator)
kafka_1               | [2020-04-30 13:55:42,781] INFO [GroupCoordinator 1001]: Assignment received from leader for group group1 for generation 1 (kafka.coordinator.group.GroupCoordinator)
```

In zookeeper logs:
```
 Error Path:/brokers Error:KeeperErrorCode = NoNode for /brokers
zookeeper_1           | 2020-04-30 13:53:14,721 [myid:] - INFO  [ProcessThread(sid:0 cport:2181)::PrepRequestProcessor@653] - Got user-level KeeperException when processing sessionid:0x10026398b3c0000 type:create cxid:0x6 zxid:0x7 txntype:-1 reqpath:n/a Error Path:/config Error:KeeperErrorCode = NoNode for /config
zookeeper_1           | 2020-04-30 13:53:14,726 [myid:] - INFO  [ProcessThread(sid:0 cport:2181)::PrepRequestProcessor@653] - Got user-level KeeperException when processing sessionid:0x10026398b3c0000 type:create cxid:0x9 zxid:0xa txntype:-1 reqpath:n/a Error Path:/admin Error:KeeperErrorCode = NoNode for /admin
zookeeper_1           | 2020-04-30 13:53:15,016 [myid:] - INFO  [ProcessThread(sid:0 cport:2181)::PrepRequestProcessor@653] - Got user-level KeeperException when processing sessionid:0x10026398b3c0000 type:create cxid:0x15 zxid:0x15 txntype:-1 reqpath:n/a Error Path:/cluster Error:KeeperErrorCode = NoNode for /cluster
zookeeper_1           | 2020-04-30 13:53:16,339 [myid:] - INFO  [ProcessThread(sid:0 cport:2181)::PrepRequestProcessor@596] - Got user-level KeeperException when processing sessionid:0x10026398b3c0000 type:multi cxid:0x38 zxid:0x1d txntype:-1 reqpath:n/a aborting remaining multi ops. Error Path:/admin/preferred_replica_election Error:KeeperErrorCode = NoNode for /admin/preferred_replica_election
zookeeper_1           | 2020-04-30 13:55:36,782 [myid:] - INFO  [ProcessThread(sid:0 cport:2181)::PrepRequestProcessor@653] - Got user-level KeeperException when processing sessionid:0x10026398b3c0000 type:setData cxid:0x40 zxid:0x1e txntype:-1 reqpath:n/a Error Path:/config/topics/__consumer_offsets Error:KeeperErrorCode = NoNode for /config/topics/__consumer_offsets
zookeeper_1           | 2020-04-30 13:55:37,766 [myid:] - INFO  [ProcessThread(sid:0 cport:2181)::PrepRequestProcessor@653] - Got user-level KeeperException when processing sessionid:0x10026398b3c0000 type:setData cxid:0xe0 zxid:0x86 txntype:-1 reqpath:n/a Error Path:/config/topics/event_topic Error:KeeperErrorCode = NoNode for /config/topics/event_topic
```

But data is not in the table engine=Kafka and in the MV

**Comment by filimonov** [comment]
Hi @radist101, clickhouse-side log may have the answers - check that: https://github.com/ClickHouse/ClickHouse/issues/9334#issuecomment-590281028

From my perspective, the fixed IP in `KAFKA_ADVERTISED_HOST_NAME` may be potential reason.

Check also that compose file: https://github.com/filimonov/ClickHouse/blob/kafka_cluster_tests/dbms/tests/integration/test_kafka_cluster/docker-compose.yml

**Comment by filimonov** [comment]
Closing due to lack of activity, and because the thread goes away from the original topic. 

If you still face the issue, please open a new one with the needed details.

**Comment by Java-xian-fish** [comment]
I have the same issue,and I have resolve this problem,change you column name from `updateTime` to `update_time`,maybe its a bug in Parsing

**Comment by xbj137661413** [comment]
make sure whether consume read_commit and
mabe you can try:
./kafka-transactions.sh --bootstrap-server {your kafka brokers}:9092 find-hanging find-hanging --topic {your topic}
and fix it:
./kafka-transactions.sh --bootstrap-server {your kafka brokers}:9092 abort --topic {your topic} --partition {hanging partition} --start-offset {start offset}
